server.go

```diff
@@ -24,113 +26,140 @@ type Route struct {
+			if errors.As(err, &errStatusCode) {
+				statusCode = int(errStatusCode)
+			}
+
+			log.Printf("Error: %v", err)
+			w.WriteHeader(statusCode)
+			return
+		}
 
 		elapsed := time.Since(start)
 
 		log.Printf("%s %s %d ms\n", r.Method, r.RequestURI, elapsed.Milliseconds())
 	})
 
 	log.Printf("INFO: Listening on port 8080\n")
-	http.ListenAndServe(":8080", nil)
+	if err := http.ListenAndServe(":8080", nil); err != http.ErrServerClosed {
+		return fmt.Errorf("HTTP Server failed: %w", err)
+	}
+
+	return nil
 }
 
-func readAndUnmarshalRoutes(routeFile string) Routes {
+func readAndUnmarshalRoutes(routeFile string) (Routes, error) {
 	data, err := os.ReadFile(routeFile)
 	if err != nil {
-		log.Fatalf("ERROR: Unable to read file: %s", err.Error())
+		return Routes{}, fmt.Errorf("unable to read file: %w", err)
 	}
 
 	var routes Routes
 	err = yaml.Unmarshal(data, &routes)
 	if err != nil {
-		log.Fatalf("ERROR: Unable to decode YAML: %s", err.Error())
+		return Routes{}, fmt.Errorf("unable to decode YAML: %w", err)
 	}
 
-	sort.Sort(ByPathLength(routes.Routes))
+	sort.Sort(byPathLength(routes.Routes))
 
-	return routes
+	return routes, nil
 }
 
-func findRoute(r *http.Request, routes Routes) *Route {
+func findRoute(r *http.Request, routes Routes) (*Route, bool) {
 	for _, route := range routes.Routes {
 		params := parseURLParameters(r.URL.Path, route.Path)
 		if params != nil && r.Method == route.Method {
 			route.Params = params
-			return &route
+			return &route, true
 		}
 	}
 
-	return nil
+	return nil, false
 }
 
-func handleHTTPRequest(w http.ResponseWriter, r *http.Request, routes Routes, cgi_scripts_dir string) {
-	route := findRoute(r, routes)
-	if route == nil {
-		http.Error(w, "no script found for route", http.StatusBadRequest)
-		return
+func handleHTTPRequest(w http.ResponseWriter, r *http.Request, routes Routes, cgiScriptsDir string) error {
+	route, ok := findRoute(r, routes)
+	if !ok {
+		return errors.Join(
+			fmt.Errorf("no script found for route"),
+			statusCodeError(http.StatusBadRequest),
+		)
 	}
 
-	statusCode, err := executeScript(w, r, *route, path.Join(cgi_scripts_dir, route.Binary))
-	if err != nil {
-		http.Error(w, err.Error(), statusCode)
+	if err := executeScript(w, r, *route, path.Join(cgiScriptsDir, route.Binary)); err != nil {
+		return fmt.Errorf("execute script: %w", err)
 	}
+
+	return nil
 }
 
-func executeScript(w http.ResponseWriter, r *http.Request, route Route, binary_path string) (int, error) {
+type statusCodeError int
+
+func (err statusCodeError) Error() string {
+	return fmt.Sprintf("status code: %d", err)
+}
 
-	ctx, cancel := context.WithTimeout(context.Background(), TIMEOUT_MILLISECONDS*time.Millisecond)
+func executeScript(w http.ResponseWriter, r *http.Request, route Route, binaryPath string) error {
+	ctx, cancel := context.WithTimeout(r.Context(), timeout)
 	defer cancel()
 
-	cmd := exec.CommandContext(ctx, binary_path)
+	cmd := exec.CommandContext(ctx, binaryPath)
```

Comment:
I have no experience with the exec package. But the documentation says, that `exec.CommandContext` will stop the process, if the context get done. So there is no need to manually handle that.

There was a timing bug in your code. When there were many parallel requests, the context could timeout before the process had even started. In this case `cmd.Process` in your line 166 was nil and calling `cmd.Process.Kill()` resulted in a panic.